crates/autopilot/src/domain/fee/policy.rs

```diff
@@ -0,0 +1,92 @@
+//! Protocol fee implementation.
+//!
+//! The protocol fee is a fee that is defined by the protocol and for each order
+//! we define the way to calculate the protocol fee based on the configuration
+//! parameters.
+
+use crate::{
+    boundary::{self},
+    domain,
+    infra::{self},
+};
+
+/// Constructs fee policies based on the current configuration.
+#[derive(Debug)]
+pub struct Policies {
+    policy: Policy,
+    fee_policy_skip_market_orders: bool,
+}
+
+impl Policies {
+    pub fn new(policy: Policy, fee_policy_skip_market_orders: bool) -> Self {
+        Self {
+            policy,
+            fee_policy_skip_market_orders,
+        }
+    }
+
+    /// Get policies for order.
+    pub fn get(&self, order: &boundary::Order, quote: Option<&domain::Quote>) -> Vec<Policy> {
```

Comment:
Does that mean that this function also does conversion from `boundary::Order` to `domain::Order`?

Reply:
I think so, although really we should always have a plan for how to phase out boundary (if I understand correctly this  module is supposed to be temporary and a tool to keep the scope of the refactoring small).

We load raw order data from the database (without fee policies). These need to be represented in the domain somehow (e.g. Orders) and shouldn't stay `boundary::Order` forever. Then it is IMO part of the domain logic to add fees to this type, which likely should yield a new domain type (PreparedOrders or something like this), which are then type used on `domain::Auction`.

Turning a Oder into a PreparedOrder is to me what the domain::Fee component should do (until we have other prep work that needs to be done by the autopilot).

One could potentially also take `domain::Order` instance with fee_policy being `[]` and mutate them here, but this could easily lead to us believing we have a "prepared auction" whereas we missed an important step of the preparation (adding fees)

Reply:
To me, we should introduce new state of the order (in your case `PreparedOrder`) only if, for example, somewhere in the middle of the autopilot domain order lifecycle, the state of the order changes (think state machine). For now, I see this as potentially needed only for `order.executed` field, but not even for this, since the state is saved to database after each trade.

IMO, what you want should be part of refactoring the solvable orders. So, there, we should have a different implementation of fetching the data from the DB (instead of OPEN_ORDERS) and with that, we should define a new dto (as a replacement for `model::Order`) and use this dto to convert to domain::Order (and that conversion would live in `infra`)

Reply:
> we should define a new dto (as a replacement for model::Order) and use this dto to convert to domain::Order (and that conversion would live in infra)

Would `domain::Order` in this world be without the fee? Applying fee logic to me is a core part of the autopilot domain and should thus not live in infra. The reason why I wouldn't want to reuse the same model (say db returns `Order` with `fee: vec![]` and then the fee domain component amends it), is that it's not clear from the type if an object is "ready" to be passed along to e.g. the `request_solve` method.